flink-runtime/src/main/java/org/apache/flink/runtime/metrics/MetricRegistry.java

```diff
@@ -216,17 +244,20 @@ public ScopeFormats getScopeFormats() {
 	 * @param metricName  the name of the metric
 	 * @param group       the group that contains the metric
 	 */
-	public void register(Metric metric, String metricName, MetricGroup group) {
+	public void register(Metric metric, String metricName, AbstractMetricGroup group) {
 		try {
 			if (reporters != null) {
-				for (MetricReporter reporter : reporters) {
+				for (int i = 0; i < reporters.size(); i++) {
+					MetricReporter reporter = reporters.get(i);
 					if (reporter != null) {
-						reporter.notifyOfAddedMetric(metric, metricName, group);
+						FrontMetricGroup front = new FrontMetricGroup(i);
```

Comment:
we will now create a new object every time we add a new metric. Instead, keep a single `FrontMetricGroup` instance in the registry, and add a `setIndex()` method to it that you call here along with `setReference()`.

but, i just realized that without this you would end up with concurrency issues since multiple register calls can be active at the same time...I'll have to think about this for a bit.